README.md
![npm version](https://img.shields.io/npm/v/spi-node/latest?style=for-the-badge)
![npm version](https://img.shields.io/npm/types/spi-node?style=for-the-badge)
![npm version](https://img.shields.io/librariesio/release/npm/spi-node?style=for-the-badge&logoColor=white)
![npm version](https://img.shields.io/node/v/spi-node?style=for-the-badge&logo=node.js&logoColor=white)

# spi-node

Communicate with devices using [SPI](https://en.wikipedia.org/wiki/Serial_Peripheral_Interface) directly from your Node.js app. Data transfers are executed asynchronously in their own thread. Supports mock data transfers on machines without SPI support for easy local development. Uses Node's now-stable [N-API](https://nodejs.org/dist/latest-v10.x/docs/api/n-api.html#n_api_n_api) for native addon support. Written in Typescript.

## Installation

To install:

```
npm install --save spi-node
```

When installing, the native binaries will be rebuilt and added to your project's `/build` directory. If installing on a machine without SPI support, you'll see this warning from the compiler:

```
warning: "SPI not supported on this machine" [-W#warnings]
```

Not a problem, you can programatically opt-in to using mock data transfers by overriding the underlying transfer function.

## Usage

[TODO: Add MCP3008 example]

# API

The `spi-node` package exports the following:

```
import {
  Flags,
  Mode,
  Order,
  SPI,
  Settings,         // TS type
  SpiAddon          // TS type
  TransferCallback, // TS type
  TransferConfig,   // TS type
  TransferFunction, // TS type
} from 'spi-node';
```

## `Flags` Enum

`Flags` is an enum of the underlying [flags used to set an SPI channel's](https://en.wikipedia.org/wiki/Serial_Peripheral_Interface#Clock_polarity_and_phase) mode. You (probably) won't need to use `Flags` directly as they're built into the values of `Mode`.

```
enum Flags {
  CPHA = 0x01,
  CPOL = 0x02
}
```

- `Flags.CPOL`: If set, sets clock polarity such that the leading edge of the clock cycle is a falling edge.
- `Flags.CPHA`: If set, data is read on the leading edge of the clock cycle.

## `Mode` Enum

`Mode` is an enum of the [four possible SPI modes](https://en.wikipedia.org/wiki/Serial_Peripheral_Interface#Mode_numbers), which are just permutations of the `Flags`.

```
enum Mode {
  M0 =	0 | 0,
  M1 =	0 | Flags.CPHA,
  M2 =	Flags.CPOL | 0,
  M3 =	Flags.CPOL | Flags.CPHA,
}
```

> **The default `mode` of an `SPI` instance is `Mode.M0`.**

## `Order` Enum

`Order` is an enum whose values set the bit order of data transferred over an SPI connection.

```
enum Order {
  MSB_FIRST = 0,
  LSB_FIRST = 1,
}
```

> **The default `order` of an `SPI` instance is `Order.MSB_FIRST`.**

## `SPI` Class

[TODO: Add descriptions]

### Static properties & constructors

##### Property `SPI.spiSupported: boolean`

...

##### Constructor `SPI.fromDevicePath(path: string)`

...

##### Constructor `SPI.fromFileDescriptor(fd: number)`

...

---

### Properties

##### `mode: Mode`

...

##### `order: Order`

...

##### `speed: number`

...

##### `transferOverride: TransferFunction | null`

...

---

### Methods

##### `write(dataIn: Buffer): Promise<Buffer>`

...

##### `read(readcount: number): Promise<Buffer>`

...

##### `transfer(dataIn: Buffer, readcount: number): Promise<Buffer>`

...


##### `close(): void`

...

---

### Chainable setter methods

All of the properties of an `SPI` instance can be set using chainable setter methods, allowing you to instantiate and configure an `SPI` connection in one go:

```
const spi = SPI.fromDevicePath('/dev/spidev0.0')
  .setMode(Mode.M0)
  .setOrder(Order.MSB_FIRST)
  .setSpeed(1e7)
  .setTransferOverride(dataIn => Promise.resolve(dataIn)); // jumper MISO/MOSI
```


CHANGELOG.md
# Changelog

## [3.0.0] - 2021-01-30

Upgrades all dependencies.

### BREAKING CHANGES

- `node-addon-api` has been upgraded to version 3, which only supports Node version 10 and above.

---

## [1.0.0] - 2018-11-19

Initial release.


test/index.js
// eslint-disable-next-line @typescript-eslint/no-var-requires
const { SPI } = require('../dist');

const spi = new SPI();

console.log(SPI);
console.log(spi);

console.log(spi.clockSpeed);
spi.clockSpeed = 1;
console.log(spi.clockSpeed);
spi.clockSpeed = -2;
console.log(spi.clockSpeed);


src/spi.ts
import * as fs from 'fs';
import { Mode, Order, SpiAddon, Settings, TransferFunction } from './spi.types';
import { constraints } from './utils';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const SpiAddon: SpiAddon = require('bindings')('spi');

export class SPI implements Settings {
  static readonly spiSupported = SpiAddon.spiSupported;

  static fromDevicePath(devicePath = '/dev/spidev0.0') {
    const fd = fs.openSync(devicePath, 'r+');
    return new SPI(fd);
  }

  static fromFileDescriptor(fd: number) {
    return new SPI(fd);
  }

  @constraints<Mode>((x) => Object.values(Mode).includes(x))
  mode: Mode = Mode.M0;

  setMode(mode: Mode) {
    this.mode = mode;
    return this;
  }

  @constraints<number>(Number.isSafeInteger, (x) => x > 0)
  speed = 4e6;

  setSpeed(speed: number) {
    this.speed = speed;
    return this;
  }

  @constraints<Order>((x) => Object.values(Order).includes(x))
  order = Order.MSB_FIRST;

  setOrder(order: Order) {
    this.order = order;
    return this;
  }

  @constraints<TransferFunction | null>((x) => x === null || typeof x === 'function')
  transferOverride: TransferFunction | null = null;

  setTransferOverride(txFn: TransferFunction) {
    this.transferOverride = txFn;
    return this;
  }

  private constructor(readonly fd: number) {}

  write(buffer: Buffer) {
    return this.transfer(buffer, 0);
  }

  read(readcount: number) {
    return this.transfer(Buffer.alloc(0), readcount);
  }

  transfer(dataIn: Buffer, readcount: number = dataIn.length): Promise<Buffer> {
    if (this.transferOverride) {
      return this.transferOverride(dataIn, readcount);
    } else {
      const config = {
        speed: this.speed,
        mode: this.mode,
        order: this.order,
        fd: this.fd,
        dataIn,
        readcount,
      };

      return new Promise<Buffer>((ok, err) =>
        SpiAddon.transfer(
          (error, dataOut) => (error ? err(error) : ok(dataOut ?? Buffer.of())),
          config,
        ),
      );
    }
  }

  close() {
    fs.closeSync(this.fd);
  }
}


src/utils.cc
#include <napi.h>
#include <sstream>

namespace NapiUtils {
	static Napi::Value getProp(Napi::Env& env, Napi::Object& obj, const char *key) {
		if (!obj.Has(key)) {
			std::stringstream ss;
			ss << "SPI config is missing the \"" << key << "\" property." << std::endl;
			throw Napi::Error::New(env, Napi::String::New(env, ss.str()));
		}

		return obj.Get(key);
	}

	template <typename T>
	static Napi::Buffer<T> getBufferProp(Napi::Env& env, Napi::Object& obj, const char *key) {
		Napi::Value _value = getProp(env, obj, key);

		if (!_value.IsBuffer()) {
			std::stringstream ss;
			ss << "SPI config property \"" << key << "\" must be a buffer." << std::endl;
			throw Napi::Error::New(env, Napi::String::New(env, ss.str()));
		}

		return _value.As<Napi::Buffer<T>>();
	}

	static Napi::Number getNumberProp(Napi::Env& env, Napi::Object& obj, const char *key) {
		Napi::Value _value = getProp(env, obj, key);

		if (!_value.IsNumber()) {
			std::stringstream ss;
			ss << "SPI config property \"" << key << "\" must be a number." << std::endl;
			throw Napi::Error::New(env, Napi::String::New(env, ss.str()));
		}

		return _value.As<Napi::Number>();
	}
}


src/spi.cc
#include <cstdlib>
#include <algorithm>
#include <iostream>
#include <sstream>
#include <stdexcept>

#include <napi.h>
#include "napi-utils.cc"

#if __linux__
#include <sys/ioctl.h>
#include <linux/spi/spidev.h>
#endif

#ifndef SPI_IOC_MESSAGE
	// Disable some diagnostics and issue a compiler warning
	#pragma GCC diagnostic ignored "-Wunused-private-field"
	#pragma GCC diagnostic ignored "-Wunused-variable"
	#ifdef __GNUC__
		#warning "SPI not supported on this machine"
	#elif
		#pragma message("SPI not supported on this machine")
	#endif

	// Define an SPI support flag
	#define SPI_SUPPORTED false

	// Copied from https://github.com/torvalds/linux/blob/master/include/uapi/linux/spi/spidev.h
	#define SPI_CPHA		0x01
	#define SPI_CPOL		0x02

	#define SPI_MODE_0		(0|0)
	#define SPI_MODE_1		(0|SPI_CPHA)
	#define SPI_MODE_2		(SPI_CPOL|0)
	#define SPI_MODE_3		(SPI_CPOL|SPI_CPHA)
#else
	// Define an SPI support flag
	#define SPI_SUPPORTED true
#endif

class SpiTransfer : public Napi::AsyncWorker {
private:
	int fd;
	uint32_t speed;
	uint8_t mode;
	uint8_t order;
	uint32_t readcount;
	size_t buflen;
	uint8_t* buffer;

public:
	SpiTransfer(
		Napi::Function& cb,
		int fd,
		uint32_t speed,
		uint8_t mode,
		uint8_t order,
		Napi::Buffer<uint8_t> dataIn,
		size_t readcount
	): 	AsyncWorker(cb),
		fd(fd),
		speed(speed),
		mode(mode),
		order(order),
		readcount(readcount)
	{
		size_t writecount = dataIn.Length();
		buflen = std::max(writecount, readcount);
		buffer = (uint8_t*) malloc(buflen);
		memcpy(buffer, dataIn.Data(), writecount);
		memset(buffer + writecount, 0, readcount >= writecount ? readcount - writecount : 0);
	}

	~SpiTransfer() {
		free(buffer);
	}

	void Execute() {
	#ifdef SPI_IOC_MESSAGE
		if (ioctl(fd, SPI_IOC_WR_MODE, &mode) == -1) {
			throw std::runtime_error("Failed to set the SPI mode.");
		}

		if (ioctl(fd, SPI_IOC_WR_LSB_FIRST, &order) == -1) {
			throw std::runtime_error("Failed to set the SPI bit order.");
		}

		spi_ioc_transfer msg({
			.tx_buf = (uintptr_t)buffer,
			.rx_buf = (uintptr_t)buffer,
			.len = buflen,
			.speed_hz = speed,
		});

		if (ioctl(fd, SPI_IOC_MESSAGE(1), &msg) == -1) {
			throw std::runtime_error("SPI message transfer failed.");
		}
	#else
		throw std::runtime_error("SPI is not supported on this machine.");
	#endif
	}

	void OnOK() {
		Napi::HandleScope scope(Env());
		Napi::Buffer<u_int8_t> dataOut = Napi::Buffer<uint8_t>::New(scope.Env(), buffer, buflen);
		Callback().Call({ scope.Env().Null(), dataOut });
	}

	void OnError(const Napi::Error& e) {
		Napi::HandleScope scope(Env());
		Callback().Call({ e.Value(), Env().Null() });
	}
};

Napi::Object ReadSpiSettings(const Napi::CallbackInfo& info) {
	auto env = info.Env();
	if (!info[0].IsNumber()) throw Napi::Error::New(env, "readSpiSettings expects a file descriptor as its argument!");
	auto fd = info[0].As<Napi::Number>().Uint32Value();
	auto settings = Napi::Object::New(env);
	settings.Set("mode", env.Null());
	settings.Set("order", env.Null());
	settings.Set("bitsPerWord", env.Null());
	settings.Set("speed", env.Null());

	#ifdef SPI_IOC_MESSAGE
		uint8_t order, bits;
		uint32_t mode, speed;
		ioctl(fd, SPI_IOC_RD_MODE32, &mode);
		ioctl(fd, SPI_IOC_RD_LSB_FIRST, &order);
		ioctl(fd, SPI_IOC_RD_BITS_PER_WORD, &bits);
		ioctl(fd, SPI_IOC_RD_MAX_SPEED_HZ, &speed);
		settings.Set("mode", Napi::Number::New(env, mode));
		settings.Set("order", Napi::Number::New(env, order));
		settings.Set("bitsPerWord", Napi::Number::New(env, bits));
		settings.Set("speed", Napi::Number::New(env, speed));
	#endif

	return settings;
}

Napi::Value Transfer(const Napi::CallbackInfo& info) {
	Napi::Env env = info.Env();

	// Check the arguments and their types
	if (!info[0].IsFunction()) throw Napi::Error::New(env, "The first argument of the SPI transfer method must be a callback function!");
	if (!info[1].IsObject()) throw Napi::Error::New(env, "The second argument of the SPI transfer method must be a config object!");
	auto cb = info[0].As<Napi::Function>();
	auto config = info[1].As<Napi::Object>();

	// Validate the config properties and initialize a transfer worker
	SpiTransfer *worker = new SpiTransfer(
	/* cb */		cb,
	/* fd */		NapiUtils::getNumberProp(env, config, "fd").Int32Value(),
	/* speed */		NapiUtils::getNumberProp(env, config, "speed").Uint32Value(),
	/* mode */		(uint8_t) NapiUtils::getNumberProp(env, config, "mode").Uint32Value(),
	/* order */		(uint8_t) NapiUtils::getNumberProp(env, config, "order").Uint32Value(),
	/* buffer */	NapiUtils::getBufferProp<uint8_t>(env, config, "dataIn"),
	/* readcount */	(size_t) NapiUtils::getNumberProp(env, config, "readcount").Uint32Value()
	);
	worker->Queue();
	return env.Undefined();
}


Napi::Object Init(Napi::Env env, Napi::Object exports) {
	Napi::Object modes = Napi::Object::New(env);
	modes.Set("SPI_MODE_0", SPI_MODE_0);
	modes.Set("SPI_MODE_1", SPI_MODE_1);
	modes.Set("SPI_MODE_2", SPI_MODE_2);
	modes.Set("SPI_MODE_3", SPI_MODE_3);
	exports.Set("modes", modes);
	exports.Set("spiSupported", Napi::Boolean::New(env, SPI_SUPPORTED));
	exports.Set("transfer", Napi::Function::New(env, Transfer));
	exports.Set("readSpiSettings", Napi::Function::New(env, ReadSpiSettings));
	return exports;
}

NODE_API_MODULE(spi, Init)
